### PR TITLE
Adds .NET quickstarts

### PR DESCRIPTION
**Pull request by nbarbettini**
## Description:

Adds quickstarts for ASP.NET Core 2.0 and ASP.NET 4.x

Note: Currently punting on the ASP.NET 4.x Code Flow example. It's unreasonably large/difficult since the tech is older and doesn't support OIDC natively. I'll circle back around to this.

## Reviewers:

@jmelberg-okta @robertjd

### CODE REVIEW

**Comment by nbarbettini** [comment]
Ready for review. Need a sanity check that I named everything right, it builds on macOS, etc. 😄

File: _source/quickstart/dotnet/aspnet4-implicit.md [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+
+Add the following code to your `Configuration` method:
+
+```csharp
+public void Configuration(IAppBuilder app)
+{
+    // Configure JWT Bearer middleware
+    // with an OpenID Connect Authority
+    
+    var authority = "https://{{your_Okta_domain}}.com/oauth2/default";
```
Comment (jmelberg-okta) [inline _source/quickstart/dotnet/aspnet4-implicit.md]: nit: `https://{{your_Okta_domain}}.com/` -> `https://{yourOktaDomain}.com/`

File: _source/quickstart/dotnet/aspnet4-implicit.md [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+
+    var configurationManager = new ConfigurationManager<OpenIdConnectConfiguration>(
+        authority + "/.well-known/oauth-authorization-server",
+        new OpenIdConnectConfigurationRetriever(),
```
Comment (jmelberg-okta) [inline _source/quickstart/dotnet/aspnet4-implicit.md]: This is a little  confusing. We are using the `oauth-authorization-server` metadata endpoint here, but calling an `OpenIDConnectConfigurationRetriever` method. We also state above that this is configured `with an OpenID Connect Authority`.

Curious if there is a generic `ConfigurationRetriever`? Or -- can we just use the OpenID Connect endpoint instead?

File: _source/quickstart/dotnet/aspnet4-implicit.md [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+        new HttpDocumentRetriever());
+
+    app.UseJwtBearerAuthentication(new JwtBearerAuthenticationOptions
+    {
+        AuthenticationMode = AuthenticationMode.Active,
+        TokenValidationParameters = new TokenValidationParameters
+        {
+            ValidAudience = "api://default",
+            ValidIssuer = authority,
+            IssuerSigningKeyResolver = (token, securityToken, identifier, parameters) =>
```
Comment (jmelberg-okta) [inline _source/quickstart/dotnet/aspnet4-implicit.md]: Where are these values defined?

File: _source/quickstart/dotnet/aspnet4-implicit.md [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+            {
+                var discoveryDocument = Task.Run(() => configurationManager.GetConfigurationAsync()).GetAwaiter().GetResult();
+                return discoveryDocument.SigningKeys;
+            }
+        }
+    });
+}
+```
```
Comment (jmelberg-okta) [inline _source/quickstart/dotnet/aspnet4-implicit.md]: Can we add a note about validating the `state` and `nonce` (if an `id_token`)

File: _source/quickstart/dotnet/aspnetcore-auth-code.md [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+    sharedOptions.DefaultAuthenticateScheme = CookieAuthenticationDefaults.AuthenticationScheme;
+    sharedOptions.DefaultSignInScheme = CookieAuthenticationDefaults.AuthenticationScheme;
+    sharedOptions.DefaultChallengeScheme = OpenIdConnectDefaults.AuthenticationScheme;
+})
+.AddCookie()
+.AddOpenIdConnect(options =>
+{
+    options.ClientId = "{{app_client_id}}";
+    options.ClientSecret = "{{app_client_secret}}";
+    options.Authority = "https://{{your_Okta_domain}}.com/oauth2/default";
```
Comment (jmelberg-okta) [inline _source/quickstart/dotnet/aspnetcore-auth-code.md]: nit: `https://{{your_Okta_domain}}.com` -> `https://{yourOktaDomain}.com`

File: _source/quickstart/dotnet/aspnetcore-implicit.md [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+In the `ConfigureServices` method, add this `UseAuthentication` block and configure it using the information from the Okta application you just created:
+
+```csharp
+services.AddAuthentication(sharedOptions =>
+{
+    sharedOptions.DefaultAuthenticateScheme = JwtBearerDefaults.AuthenticationScheme;
+    sharedOptions.DefaultChallengeScheme = JwtBearerDefaults.AuthenticationScheme;
+})
+.AddJwtBearer(options => {
+    options.Authority = "https://{{your_Okta_domain}}.com/oauth2/default";
```
Comment (jmelberg-okta) [inline _source/quickstart/dotnet/aspnetcore-implicit.md]: nit: `https://{{your_Okta_domain}}.com` -> `https://{yourOktaDomain}.com`

File: _source/quickstart/dotnet/aspnet4-implicit.md (hunk 4 above)
Comment (nbarbettini) [reply _source/quickstart/dotnet/aspnet4-implicit.md]: Would a SPA send an `id_token` to an API? What's the use case?

File: _source/quickstart/dotnet/aspnet4-implicit.md (hunk 3 above)
Comment (nbarbettini) [reply _source/quickstart/dotnet/aspnet4-implicit.md]: They're not, it's a set of named lambda parameters (note the `=>`)

File: _source/quickstart/dotnet/aspnet4-implicit.md (hunk 1 above)
Comment (nbarbettini) [reply _source/quickstart/dotnet/aspnet4-implicit.md]: Curious, why is this important?

Comment (jmelberg-okta) [reply _source/quickstart/dotnet/aspnet4-implicit.md]: Not super important. But we use camelCase everywhere else.

File: _source/quickstart/dotnet/aspnet4-implicit.md (hunk 4 above)
Comment (jmelberg-okta) [reply _source/quickstart/dotnet/aspnet4-implicit.md]: Ah, fair point. Neither of those validation params are needed for the implicit use case.

File: _source/quickstart/dotnet/aspnet4-implicit.md (hunk 2 above)
Comment (nbarbettini) [reply _source/quickstart/dotnet/aspnet4-implicit.md]: Good catch - I pasted the wrong well-known path.

**Comment by nbarbettini** [comment]
@jmelberg-okta Rebased and changed PR base to `jm-138486-update-docs-index`. Is that what you meant? If so, feel free to merge.

File: _source/quickstart/dotnet/aspnet4-implicit.md [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+
+Add the following code to your `Configuration` method:
+
+```csharp
+public void Configuration(IAppBuilder app)
+{
+    // Configure JWT Bearer middleware
+    // with an OpenID Connect Authority
+    
+    var authority = "https://{{yourOktaDomain}}.com/oauth2/default";
```
Comment (bdemers) [inline _source/quickstart/dotnet/aspnet4-implicit.md]: remove duplicate brackets this renders as missing property (empty string)
`"https://.com/oauth2/default"`
(each file)

File: _source/quickstart/dotnet/aspnetcore-auth-code.md [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+        {
+            return SignOut(CookieAuthenticationDefaults.AuthenticationScheme, OpenIdConnectDefaults.AuthenticationScheme);
+        }
+        
+        return RedirectToAction("Index", "Home");
+    }
+}
+```
+
+> Note: If you want to log the user out, you'll need to add a **Logout redirect URI** to the application configuration in Okta with a value of `http://{{your_server_URL}}/signout-callback-oidc`. For example, `http://localhost:5000/signout-callback-oidc`.
```
Comment (bdemers) [inline _source/quickstart/dotnet/aspnetcore-auth-code.md]: extra brackets

File: _source/quickstart/dotnet/aspnetcore-auth-code.md [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+{
+    sharedOptions.DefaultAuthenticateScheme = CookieAuthenticationDefaults.AuthenticationScheme;
+    sharedOptions.DefaultSignInScheme = CookieAuthenticationDefaults.AuthenticationScheme;
+    sharedOptions.DefaultChallengeScheme = OpenIdConnectDefaults.AuthenticationScheme;
+})
+.AddCookie()
+.AddOpenIdConnect(options =>
+{
+    options.ClientId = "{{clientId}}";
+    options.ClientSecret = "{{clientSecret}}";
```
Comment (bdemers) [inline _source/quickstart/dotnet/aspnetcore-auth-code.md]: Brackets, brackets :P

**Comment by bdemers** [comment]
I was able to get through the `aspnetcore-auth-code.md` with out an issue on a Mac using Intellij's Rider (and the command line)

**Comment by nbarbettini** [comment]
Thank you for the UAT @bdemers!

File: _source/quickstart/dotnet/aspnet4-implicit.md (hunk 7 above)
Comment (nbarbettini) [reply _source/quickstart/dotnet/aspnet4-implicit.md]: To clarify: you're saying it should be `{yourOktaDomain}`?

File: _source/quickstart/dotnet/aspnetcore-implicit.md [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+```csharp
+[Authorize]
+[Route("/api")]
+public class MessagesController : Controller
+{
+    [HttpGet]
+    [Route("messages")]
+    public IEnumerable<string> GetPrivateMessages()
+    {
+        var login = HttpContext.User.Claims
```
Comment (bdemers) [inline _source/quickstart/dotnet/aspnetcore-implicit.md]: I had to use `var login = User.Claims`. I don't know why, but after this change I was able to get an implicit flow working

File: _source/quickstart/dotnet/aspnet4-implicit.md (hunk 7 above)
Comment (bdemers) [reply _source/quickstart/dotnet/aspnet4-implicit.md]: correct  :)

**Pull request merged**